### ISSUE TITLE
[Feature] Manifest V3 for Chrome

### ISSUE THREAD

**Issue by mknicely-atlassian** [violentmonkey/violentmonkey#1934]
**Motivation**
Chrome has deprecated support for Manifest V2. I'm in a corporate environment so we've been able to use it this year but that ends in 2023. When Chrome drops Manifest V2 support Violentmonkey won't work anymore. Side note, MS Edge won't run Manifest V2 extensions anymore starting January, 2024.

**Proposed Solution**
Modify the project to use a V3 manifest, completely replacing V2.

https://developer.chrome.com/docs/extensions/migrating/
https://extensionworkshop.com/documentation/develop/manifest-v3-migration-guide/
https://blogs.windows.com/msedgedev/2022/12/05/manifest-v3-and-microsoft-edge-add-ons/

**Use Cases**
When a browser does not support Manifest V2, Violentmonkey can still be installed and used.

**Comment by tophf** [comment]
ManifestV3 doesn't provide the necessary API for userscripts yet. It's also bugged and castrated, so we can't switch.

**Comment by mknicely-atlassian** [comment]
Heh, I'm not so familiar with Github... I found the prior feature requests related to V2.

Does this mean support for latest Chrome and Edge are being dropped from the project in January?

**Comment by tophf** [comment]
Assuming they don't care about userscripts, yes.

**Comment by tophf** [comment]
Chrome team is working on a new userscripts API, so I don't think their product managers will be allowed to disable ManifestV2 for userscript extensions such as ours and Tampermonkey without giving us the necessary time to test the new API.

**Comment by tophf** [comment]
The userScripts API is ready for testing in Canary so I'll start migrating. It should improve/simplify performance for many scripts, but the API has a few nasty limitations such as the lack of `tld`, regexps in `@include`, no smart blacklist. It means I'll have to come up with possible complicated workarounds...

**Comment by tophf** [comment]
Nothing's changed, it's just as bad. We'll use the new API eventually though.

**Comment by arjpar** [comment]
I believe that Manifest v3 is possible. There are userscript managers for Safari that use MV3. I'm certain there will be a sacrifice in the power of userscripts, though.

**Comment by arash2060** [comment]
This seems to be one path to take
https://developer.chrome.com/docs/extensions/reference/api/userScripts

TamperMonkey is also still on V2. I saw a discussion here:
https://github.com/Tampermonkey/tampermonkey/issues/644

**Comment by arjpar** [comment]
@tophf MV2 will become disabled in pre-stable versions of Chrome (Dev, Canary, Beta) in June 2024. In Chromium 127 MV2 will be disabled. Current version is 124. 127 will come in 3 September 2024, so there's 4 months left to transition. [This thread](https://github.com/w3c/webextensions/issues/279) has a lot of info (which as I just checked, you have contributed to). Additionally Tampermonkey Beta [has switched](https://github.com/Tampermonkey/tampermonkey/issues/644#issuecomment-2094925643) to MV3.

**Comment by tophf** [comment]
I'm not sure we should update our stable extension because MV3 breaks some things and there's no good substitute yet, we should probably suggest the users just enable the policy to prolong ManifestV2 as [shown here for Windows](https://www.reddit.com/r/browsers/comments/119x3qo/enabling_manifestv2_and_certain_external/), we should probably describe examples for all platforms somewhere.

As for beta, I don't know. Maybe. I was planning to do it anyway this month.

**Comment by arjpar** [comment]
@tophf Why not have one extension that is MV3 and one that is MV2, like uBlock Origin & uBlock Origin Lite? How much functionality is lost in MV3?

**Comment by tophf** [comment]
I've listed some issues I remembered in Tampermonkey thread.

**Comment by cyfung1031** [comment]
> @tophf Why not have one extension that is MV3 and one that is MV2, like uBlock Origin & uBlock Origin Lite? How much functionality is lost in MV3?

Based on TM MV3 Beta, even the `GM_setValue` `GM_getValue` can be reversed without changes.
It looks like no functionality loss in the current MV3 spec can be achievable, just except the webRequest related issues to be confirmed.

**Comment by cyfung1031** [comment]
I have seen the frequent updates for VM.
Can you just frozen the current beta version to 2.19.0 and then focus on the MV3 version?

I don't think the minor updates in the current VM 2.18.x is necessary. They are just minor UI issues. It should become a stable version.

**Comment by tophf** [comment]
The problem with GM_getValue is different and it doesn't affect everyone. There are other problems I've listed.

**Comment by l1n** [comment]
Excited for this - while MV3 is really annoying in a bunch of ways, Chrome continues to make MV2 harder and harder to use, especially in compliance-heavy environments.

**Comment by jesus2099** [comment]
> @tophf https://github.com/violentmonkey/violentmonkey/issues/1934#issuecomment-2095213985
> _we should probably suggest the users to enable the policy to prolong ManifestV2 as [shown here for Windows](https://www.reddit.com/r/browsers/comments/119x3qo/enabling_manifestv2_and_certain_external/)_

This solution to keep Manifest V2, based on `HKEY_LOCAL_MACHINE` registry, requires administration rights on Windows. :-/

**Comment by tophf** [comment]
It might work for the user account if you replace HKEY_LOCAL_MACHINE with HKEY_CURRENT_USER, could you test?

**Comment by mon-jai** [comment]
https://github.com/BlackGlory/eternity is based on Manifest V3 and is working for me.

Instead of rewriting everything, perhaps we could have Violentmonkey based on it, add support for the `GM*` API, and port the current user interface?

**Comment by tophf** [comment]
We don't need to rewrite everything either, it's not the problem.

**Comment by arjpar** [comment]
@tophf Will Violentmonkey still be on MV2 when the end-user transition begins next week? I take it from this thread Violentmonkey will require users to edit the registry (Windows) or install profiles (macOS) for functionality to work once the end-user transition begins.

**Comment by tophf** [comment]
Chrome will show a button to keep the extension, so depending on how many users do that Chrome developers will decide on the availability of this button.

**Comment by arjpar** [comment]
I still think it would be a good idea to update at least the Beta extension to MV3 or create a new extension (based on Violentmonkey) that is MV3-based. While MV3 removes a lot of the power of MV2-based userscript managers, it greatly improves security and efficiency. If there isn't an ongoing attempt to port it to MV3, I may do it myself in a fork (and eventually push upstream if desired).

**Comment by tophf** [comment]
> MV3 [...] it greatly improves security and efficiency.

It doesn't, those claims are greatly exaggerated to the point of being unsubstantiated propaganda.

* Nothing changes in regards to security for users of extensions that didn't do anything insecure.
* Efficiency is lower for extensions like ours that need to process every navigation because MV3 disallows persistent background scripts, which means it will restart hundreds of times a day wasting resources much more than in MV2. No one in Chrome team has ever measured the impact, they just believe in an idealistic notion.
* MV3 userScripts API doesn't support regular expressions, so extensions have to run all such scripts in all tabs and frames, and perform the check there, which means that every regular expression will be recompiled again and again without caching, thousands of times each day. The scripts may be very big, so megabytes of code will waste the CPU time and battery without being actually used.
* MV3 userScripts doesn't support smart blacklists offered by Violentmonkey, see the previous consideration minus the regular expression compilation.
* If we don't use the very limited MV3 userScripts API, we can't ensure the scripts will run at `document_start`:
   * "synchronous page mode" option won't be available as MV3 removes webRequestBlocking 
   * the background script will need 50ms - 500ms to start and load the scripts every time navigation occurs after a pause of 30 seconds, which may easily happen hundreds of times a day.

**Comment by tophf** [comment]
I think we will make an MV3 version anyway because userScripts API offers some good features like world isolation and true document-start.

**Comment by cyfung1031** [comment]
> While MV3 removes a lot of the power of MV2-based userscript managers, 

This might not be true because the author of Tampermonkey has investigated how to use the new APIs to do the same as MV2. Only GM.webRequest is lost.  

The TM MV3 beta is already released and you can find the crx download link in TM github repo issue.

**Comment by tophf** [comment]
Tampermonkey MV3 version has all the problems I've listed.

**Comment by tophf** [comment]
But indeed, the power is practically the same, so @cyfung1031 is correct.

**Comment by railgun77** [comment]
@tophf 

> I think we will make an MV3 version anyway because userScripts API offers some good features like world isolation and true document-start.

please, I am begging you like the baggiest beggar in all of begdom 

just do not make your great extension use mv3 until absolutely necessary around the summer of 2025
for example, the developer of ublock origin plan to support the mv2 until it is impossible to do so 
and
the developer of Tampermonkey even though he update to mv3 he released a mv2 version called "Tampermonkey Legacy" because he knows that the mv3 version will cause big issues

and if you go to Tampermonkey GitHub if the script is broken he will tell you the only fix is to install "Tampermonkey Legacy"

hell if just google "Tampermonkey Legacy" you will get results of userscript authors that tell to install the mv2 version and **ignore the mv3 version**
 
and i do know one or two things about javascript and i fully understand and believe you that are some things that are good about mv3 update but for every one thing that is good about mv3 there are ten things bad about it is like one step forward and ten step backwards

and the [supermium](https://github.com/win32ss/supermium)plan to support mv2 even in 2026

**you absolutely can do whatever you want because you are the developer** 

**I just hope my begging my change your mind**

if not, then I will just grab the last mv2 version from GitHub and never update the Extension

and please ignore people like @arjpar @l1n @mknicely-atlassian 
these people just drink google Kool-Aid

**Comment by l1n** [comment]
I don't particularly enjoy that my hand is being forced towards MV3, but I can't run MV2 in an chrome enterprise environment, and I really like Violentmonkey in my personal Firefox environments.

**Comment by Xia0xia0Xia0** [comment]
# Bad news:

![screenshot_2024-07-28_20-57-05](https://github.com/user-attachments/assets/15db7910-57eb-47d8-9534-cf80a5ef9aa3)

**Comment by SergioHLX** [comment]
> just do not make your great extension use mv3 until absolutely necessary around the summer of 2025 for example

I also want the MV2 functionality to remain.  That said, google is already encouraging people not to download this because it "does not follow best practices".  Also, google may remove that functionality in the Summer of 2025, but they may begin dropping apps that depend on MV2 right away.

Possibly the thing to do is to update to MV3 now (so as to keep the ratings, metrics & userbase) & to create a new branch for the legacy for those of us who need it until google pries it from us.

**Comment by Maartyl** [comment]
I just want to make sure:  Even when Violentmonkey is forced into MV3, the Firefox version will stay unaffected?

(as the zip/xpf versions seem identical (up to signature), and from what I can understand of the build pipeline/release, there is a single version for both)
(as; Rather than switching to MV3, I am finally switching to Firefox.)

Thank you.

**Comment by Tech-How** [comment]
> I just want to make sure: Even when Violentmonkey is forced into MV3, the Firefox version will stay unaffected?
> 
> (as the zip/xpf versions seem identical (up to signature), and from what I can understand of the build pipeline/release, there is a single version for both) (as; Rather than switching to MV3, I am finally switching to Firefox.)
> 
> Thank you.

I'd like to know this as well.

**Comment by elliotberry** [comment]
I know this is the wrong forum for this, but I just want to express my extreme dissatisfaction and, candidly, unfiltered rage around Chrome continually becoming less hostile to the end user, more dumbed down, and in elbowing out any possible useful extension in favor of vapid, garbage-fire AI features that tell you to put glue on pizza.

The web has steadily become less interested in customizability and in-depth interaction, and the age-old war against userscripting reflects this.

Anyway, thanks so much, folks, for continuing to actively develop this.

**Comment by ntt77** [comment]
so, what is actually happening here?
will you develop a new version for MV3 or I need to switch to Tampermonkey?

**Comment by tophf** [comment]
Nothing's happening yet, don't have much time. Eventually we'll migrate. I'm not particularly flustered because it's quite trivial to prolong ManifestV2 until the next summer via policy/registry.

**Comment by Tech-How** [comment]
I hope we can have enough contributors to maintain both versions, or maybe somebody can fork it at some point. I don't know whether or not that would be worth it, we'll have to see. Don't want to make too many assumptions yet though, but yeah it's really sad.

**Comment by hyperion4** [comment]
@tophf
Please do not listen to people telling you not to update the existing mv2 extension with the existing id in chrome store to mv3 when you have ready the mv3 version of Vilentmonkey.
They only think about themselves and not about your great extension. If you don't do it and use a new id for the mv3 version Violentmonkey will lose its metrics, ratings and userbase. Because users will follow what the store suggests them to do and switch to Tampermonkey.
The sooner you have the mv3 ready the less users Violentmonkey will lose from Tampermonkey.
Btw I tried Tampermonkey MV3 and it is good. But I would never switch to it, to its MV2 or MV3 version. I hate a lot of stuff on it, the way it updates scripts, the nags it has. I just don't like it.
I actually believe the sooner you release Violentmonkey MV3 the better for Violentmonkey.

**Comment by hyperion4** [comment]
@gunir
I leave all the blames to Google. And?
Is that helping Violentmonkey in any way?
I will switch to Thorium (btw I am using Brave not Chrome). And?
Is that helping Violentmonkey in any way?
Stop thinking only about yourself.
Violentmonkey needs to survive and the only way is to have an always working version in  chrome store because it is the browser that has an enormous userbase.
The sooner there is a MV3 version of Violentmonkey in chrome store, the better for Violentmonkey's future and popularity because of Tampermonkey MV3.
Again, stop thinking only about yourself, MV3 won't go away just because you want it to go away.

**Comment by jleroy** [comment]
> Leave all the blames to Google, they created this MV3 disastrous, it's not better, not faster, nor more secure in fact it's nerfed in every factor.

This is simply false, but that’s not the point here.


> There's a better way, stop using Google Chrome and switch to Thorium, it'll be supporting MV2 and Violentmonkey forever

Peoples who are telling you that, whether it is the Firefox team, the Thorium one, etc. …are simply lying to you. 
Browsers needs to makes major changes like this one to move forward. 
The refusal to make these kinds of hard decisions explains (in part) why Firefox's market share has fallen so dramatically in recent years.

**Comment by jleroy** [comment]
Your « evidence » doesn't prove anything, except that you don’t know what you’re talking about.
Please stop spamming this issue and go post your rants on Google Community Forum.

**Comment by hyperion4** [comment]
> Move to what ? Do you know that to make MV3 Violentmonkey work there's a lot of hacky things that need to implement, it's more complicated than MV2 Violentmonkey for something inferior, lmao. That's the whole point of Google for deprecating MV2 lmao, and obviously, killing/nerfing adblock. Prove that MV3 is faster (maybe 1% faster on slowest hardware), more secure.

Nobody said that the MV3 version of Violentmonkey will be better than the MV2 version.
If you want to rant about Google this is not the place.
This issue is about Violentmonkey and its future in the most popular browser around.
And it happens to be owned by Google no matter if you like it or not.
And right now people are been "warned" by Google to uninstall Violentmonkey and to imstall Tampermonkey.
And they are doing it and will keep doing it until this "warning" is removed.
No, they won't install Thorium, because gunir wrote it in github lmao, they will uninstall Violentmonkey and switch to Tampermonkey.

**Comment by Tech-How** [comment]
I wish the best for the future of this project and hope that its userbase and developers remain strong and passionate. It's my favorite userscript manager at the time of this writing with an awesome design and feature set.

**Comment by lamrongol** [comment]
This is not a fundamental solution but I found **Firefox-based** and sophisticated browser(like Vivaldi) exist:
https://github.com/Floorp-Projects

**Comment by jesus2099** [comment]
Today, @Violentmonkey stopped working completely in @KiwiBrowser 124.0.6327.4, even if _developer mode_ is activated in _Extensions_ page.

https://community.metabrainz.org/t/user-scripts-on-mobile/134406/27?u=jesus2099

https://github.com/kiwibrowser/src.next/issues/715

**Comment by tophf** [comment]
Kiwi problem is not related to ManifestV3. It's reported in #2180.

**Comment by SiriosDev** [comment]
So? There Are plans for mv3 or not?

**Comment by GunGunGun** [comment]
And turning on Developer Mode is not something you can expect more than 0.001% of userbase can do

https://github.com/Tampermonkey/tampermonkey/issues/2018

MV3 Userscript is cursed.

**Comment by hyperion4** [comment]
> And turning on Developer Mode is not something you can expect more than 0.001% of userbase can do
> 
> [Tampermonkey/tampermonkey#2018](https://github.com/Tampermonkey/tampermonkey/issues/2018)
> 
> MV3 Userscript is cursed.

Tampermonkey has solved this nicely by a warning that it needs to be enabled and detailed instructions where to go and what needs to be clicked. I am sorry but I dare to say than more than 0.001% can click a button lol. It is as easy as installing an extension.

**Comment by GunGunGun** [comment]
> > And turning on Developer Mode is not something you can expect more than 0.001% of userbase can do
> > [Tampermonkey/tampermonkey#2018](https://github.com/Tampermonkey/tampermonkey/issues/2018)
> > MV3 Userscript is cursed.
> 
> Tampermonkey has solved this nicely by a warning that it needs to be enabled and detailed instructions where to go and what needs to be clicked. I am sorry but I dare to say than more than 0.001% can click a button lol. It is as easy as installing an extension.

Lamo, overestimate userbase skill is equally stupid, 96.69% of userbase are end-user, those who have absolute no ideas about checking a checkbox.

Enjoy your delusional.

**Comment by hyperion4** [comment]
> > > And turning on Developer Mode is not something you can expect more than 0.001% of userbase can do
> > > [Tampermonkey/tampermonkey#2018](https://github.com/Tampermonkey/tampermonkey/issues/2018)
> > > MV3 Userscript is cursed.
> > 
> > 
> > Tampermonkey has solved this nicely by a warning that it needs to be enabled and detailed instructions where to go and what needs to be clicked. I am sorry but I dare to say than more than 0.001% can click a button lol. It is as easy as installing an extension.
> 
> Lamo, overestimate userbase skill is equally stupid, 96.69% of userbase are end-user, those who have absolute no ideas about checking a checkbox.
> 
> Enjoy your delusional.

I am not going to reply to you the same way and calling you names, like delusional, stupid etc.
I am older than 5 years old lol.
Yes, people who use a userscript manager can click a button lmao.
Like the buttons they have to click right now for the MV2 warnings and the button they are clicking to install Tampermonkey.
Like having to apply policies in their system to postpone the inevitable.
It doesn't matter anyway, the next months they will see their userscripts not working because Violentmonkey will be disabled.
Google is going to disable Violentmonkey, they added now the ExtensionManifestV2Disabled flag.
The waring won't be there for a very long time, Violentmonkey will be disabled.
So they will go to chrome store and install Tampermonkey for having their scripts back.
No, they won't apply policies to their PC to postpone the inevitable lol.

**Comment by JnTon** [comment]
I just opened  Google Chrome (`129.0.6668.37`) and I received the notification that the extension has been automatically disabled. ![image](https://github.com/user-attachments/assets/47259889-051b-4262-b15d-b2b972afd5e9)
Additionally, I can no longer download the [stable](https://chromewebstore.google.com/detail/violentmonkey/jinjaccalgkegednnccohejagnlnfdag) and [beta](https://chromewebstore.google.com/detail/violentmonkey-beta/opokoaglpekkimldnlggpoagmjegichg) versions from the Chrome Web Store.
![image](https://github.com/user-attachments/assets/eb333f72-3cdc-49b4-acec-3eab3b0163ec)
Will a Manifest V3-compatible version be available? For now, I'm forced to switch to Tampermonkey while I wait to see if I should completely remove the Violentmonkey extension or keep it for now.

**Comment by hyperion4** [comment]
> ExtensionManifestV2Disabled

Exactly sir. Chrome won't warn forever, they will start disabling Violentmonkey.
They added the ExtensionManifestV2Disabled flag recently.
You can revert it for now by going to chrome://flags and disable ExtensionManifestV2Disabled.
But according to [GunGunGun](https://github.com/GunGunGun) this is easy and 99,99% of the userbase can do it lol.
What they can't do is to click dev mode in extensions lol.

**Comment by qzmtch** [comment]
> @gunir I leave all the blames to Google. And? Is that helping Violentmonkey in any way? I will switch to Thorium (btw I am using Brave not Chrome). And? Is that helping Violentmonkey in any way? Stop thinking only about yourself. Violentmonkey needs to survive and the only way is to have an always working version in chrome store because it is the browser that has an enormous userbase. The sooner there is a MV3 version of Violentmonkey in chrome store, the better for Violentmonkey's future and popularity because of Tampermonkey MV3. Chrome store is redirecting to it. Again, stop thinking only about yourself and what you want, MV3 won't go away just because you want it to go away.

placement in the chrome store does not give you popularity, popularity is made by people with their discussions on the Internet, when users recommend an extension to each other

**Comment by deathtrip** [comment]
When google removes support for MV2 extensions, all the recommendations and discussions on the internet won't bring Violentmonkey it's popularity back.

**Comment by qzmtch** [comment]
> When google removes support for MV2 extensions, all the recommendations and discussions on the internet won't bring Violentmonkey it's popularity back.

whatever reputation there is now, that's what it will remain, whoever uses it will continue to use it, and whoever uses tampermonkey will use it, and won't even look at violentmonkey
and discussions and recommendations on forums, social networks and in personal communication with friends will work better than being in a store

**Comment by hyperion4** [comment]
> placement in the chrome store does not give you popularity, popularity is made by people with their discussions on the Internet, when users recommend an extension to each other

That's just your wishful thinking.
If that was true then Tampermonkey which is a more heavy extension than Violentmonkey, it uses analytics etc wouldn't be more popular than Violentmonkey. People see Tampermonkey... 11.000.000 users , Violentmonkey 1.000.000 users.. so they think Tampermonkey is better because this is people use and install that.
And that will only get worse because people now see they should not install Violentmonkey because it breaks properextensions  policies... They are people out there that may believe this is a malicious extension. What is going to save Violentmonkey from that reputation? Discussions in forums? Please.

**Comment by qzmtch** [comment]
> so they think Tampermonkey is better because this is what most people use and install that.

no, that's not why, but because the creators of userscripts always write that their script works in tampermonkey, but they don't guarantee that it will work in other extensions (by the way, there are extensions that actually work in tampermonkey and don't work in violentmonkey)
it's all because of those who write scripts

**Comment by qzmtch** [comment]
> What is going to save Violentmonkey from that reputation? Discussions in forums? Please.

exactly

**Comment by hyperion4** [comment]
> > so they think Tampermonkey is better because this is what most people use and install that.
> 
> no, that's not why, but because the creators of userscripts always write that their script works in tampermonkey, but they don't guarantee that it will work in other extensions (by the way, there are extensions that actually work in tampermonkey and don't work in violentmonkey) it's all because of those who write scripts

People just go to sites like greasyfork and install scripts.
They don't even care or know who wrote them lol or care to listen what the creator recommends to use.

**Comment by qzmtch** [comment]
> People just go to sites like greasyfork and install scripts.
> They don't even care or know who wrote them lol or care to listen what the creator recommends to use.
> That's just your wishful thinking and how an ideal world should be.
> People don't care about all that.

then google is doing the right thing by prohibiting extensions from doing what it prohibits

**Comment by qzmtch** [comment]
> They don't even care or know who wrote them lol or care to listen what the creator recommends to use.

This care me and all the people I know.

**Comment by hyperion4** [comment]
> > People just go to sites like greasyfork and install scripts.
> > They don't even care or know who wrote them lol or care to listen what the creator recommends to use.
> > That's just your wishful thinking and how an ideal world should be.
> > People don't care about all that.
> 
> then google is doing the right thing by prohibiting extensions from doing what it prohibits

Of course it doesn't. Just because most people don't care that doesn't mean everybody should suffer:)

**Comment by knd775** [comment]
Stop, please. This is not what GitHub issues are for.

**Comment by hyperion4** [comment]
> Stop, please. This is not what GitHub issues are for.

You are right, I am not to reply again to this kind of discussion, you are 100% correct it is off topic.

**Comment by Purfview** [comment]
Not sure I understand, will VM support MV3 or will it be eventually bananed in that browser?
Users now just install VM instead of TM as a fix,

> google is doing the right thing

That's oxymoron joke, no?

**Comment by qzmtch** [comment]
> That's oxymoron joke, no?

in the context of the quote that was responded to, no

**Comment by cyfung1031** [comment]
> Not sure I understand, will VM support MV3 or will it be eventually bananed in that browser? 

There is no milestone / target date of MV3 version release (even alpha release) provided by Violentmonkey's authors.

**Comment by torb235** [comment]
if mv3 so great like how some **tech bros** will tell you and how mv2 is evil and needs to die 
then the Tampermonkey author would have to bother to release a "legacy" version of Extension

https://chromewebstore.google.com/detail/tampermonkey-legacy/lcmhijbkigalmkeommnijlpobloojgfn

the first adopter of mv3 superscript still intends to support mv2 until the h2 of 2025 
then that should tell about who grate the so called mv3 our lord and saviour

so i fully accepted that mv2 will die in h2 of 2025
but there is **no reason not to delay it as long as possible** 
even the Tampermonkey author gave the **option** to the user if they are interested to delay it as long as possible

**Comment by torb235** [comment]
And here is a real life example about how mv3 **breaks** some userscripts 

[this script from a forum is used to download images in batch from the forum, it stopped working because Tampermonkey updated to mv3](https://simpcity.su/threads/forum-post-downloader-tampermonkey-script.96714/page-27#post-7303407)

and the people recommended installing the mv2 version to fix the issue

after that, the author of the script stated recommending the mv2 version

**Comment by NeurekaSoftware** [comment]
Unfortunately, MV3 is here to stay. It's better to adapt to it as quickly as possible than delay the inevitable... Otherwise, the current extension will break and Violentmonkey will cease to work as there isn't an MV3 extension for it. This is the outcome we should be trying to avoid.

**Comment by qzmtch** [comment]
Firefox will have Manifest 2, there is no point in completely abandoning extensions with Manifest 2 and all the functions. We need to leave the extension as it is now and continue developing it for Firefox, and make a separate version for Chrome

**Comment by NeurekaSoftware** [comment]
> We need to leave the extension as it is now and continue developing it for Firefox, and make a second version for Chrome

This works too, but the MV3 extension shouldn't be delayed.

I'd switch to Firefox if there wasn't that annoying scrolling feature for tabs on desktop. As far as I'm aware that can't be turned off, and I need to be able to see all my tabs at a glance and not have them hidden. Dx

**Comment by hyperion4** [comment]
> Firefox will have Manifest 2, there is no point in completely abandoning extensions with Manifest 2 and all the functions. We need to leave the extension as it is now and continue developing it for Firefox, and make a second version for Chrome

tophf has released a mv3 version of Stylus.

https://github.com/openstyles/stylus/issues/1430#issuecomment-2453057677

If I understood correctly he will keep 2 completely separate versions of Stylus, the MV2 version of Stylus for Firefox and AMO and the MV3 version of Stylus for Chromium and Chrome store.

https://github.com/openstyles/stylus/issues/1430#issuecomment-2466532079

Most likely he will do the same the same for Violentmonkey if he decides to keep supporting Chromium.

**Comment by Purfview** [comment]
I would charge money for a MV3 plugin or make it with ads, those infected users would mind that at all. 😆


The developers of the userscripts, you can inform your users about what's going on: 

![xxxxxxxxxxxx](https://github.com/user-attachments/assets/c51a5450-36f5-4f3e-a191-28d0c67550ab)

**Comment by B1ack3ye** [comment]
Today, I can no longer install Violentmonkey via the Chrome Web Store. Chrome version 133.0.6943.54

![Image](https://github.com/user-attachments/assets/ff4ac1e8-f0fb-4bf7-b4ba-081799c04296)

**Comment by SekiBetu** [comment]
I'd suggest weeding out the incompatibilities, developing beta versions, and then solving the problems one by one

**Comment by Hadden89** [comment]
Tampermonkey now is MV3, but for still some months I can use Violentmonkey (tnx to the policy) :)
I think the beta would be nice to start =)

**Comment by awt-256** [comment]
User of violentmonkey for users. Chrome v133.0 just wiped the extension off my computer and I seem to have lost all my userscripts (besides ones I archived long before). For anyone who hasn't experienced this yet, take this as a time to archive your userscripts.

**Comment by SergioHLX** [comment]
> User of violentmonkey for years. Chrome v133.0 just wiped the extension off my computer and I seem to have lost all my userscripts (besides ones I archived long before). For anyone who hasn't experienced this yet, take this as a time to archive your userscripts.

Sorry that happened to you. It's frustrating & we've all been there.  That said, any setting of any Chrome extension is deleted when the extension is uninstalled. Developers who aren't using GIT, should be. Or you can turn on the built-in Sync function that Violentmonkey has so that you'll at least have the latest copy.

**Comment by dvirtz** [comment]
For me 133 just disabled the extension so I managed to import everything to tampermonkey.
So Long, and Thanks for All the Fish.

**Comment by NeurekaSoftware** [comment]
Is the official stance on this that there will not be an MV3 version and to use Firefox or a browser that continues to support MV2?

I'm only asking because I have problems with Firefox in terms of how tabs on handled on both desktop and mobile that negatively impacts my workflow, and if Violentmonkey isn't going to get updated to MV3 then I need to switch to Tampermonkey and stop recommending it to friends and family.

**Comment by NeurekaSoftware** [comment]
> guys calm down, there's still four months to go, I'm sure the developers already have a fledgling version, they just don't feel like releasing it until it reaches perfection

There's literally no indicator that work has even begin... the project is open source. It's open for everyone to see.

**Comment by hyperion4** [comment]
> guys calm down, there's still four months to go, I'm sure the developers already have a fledgling version, they just don't feel like releasing it until it reaches perfection

I don't think so. They would have done already what they did with Stylus. They made the beta version of Stylus MV3 compatible and with user feedback it will reach perfection until June.
How Violentmonkey MV3 will reach perfection without user feedback? There would be at least an alpha Violentmonkey MV3 by now if they intended to make an MV3 version.
Most likely scenario, there won't be a MV3 version of Violentmonkey.
And NeurekaSoftware is totally right, they should inform users about it so they have the time to switch and become familiar with Tampermonkey and how it works.
People are losing their userscripts because Google has started disabling Violentmonkey.
That's not cool and people should be informed by the developers of Violentmonkey in order to save them before losing them and import them to Tampermonkey.

**Comment by SekiBetu** [comment]
> > MV3 [...] it greatly improves security and efficiency.
> 
> It doesn't, those claims are greatly exaggerated to the point of being unsubstantiated propaganda.
> 
> * Nothing changes in regards to security for users of extensions that didn't do anything insecure.
> * Efficiency is lower for extensions like ours that need to process every navigation because MV3 disallows persistent background scripts, which means it will restart hundreds of times a day wasting resources much more than in MV2. No one in Chrome team has ever measured the impact, they just believe in an idealistic notion.
> * MV3 userScripts API doesn't support regular expressions, so extensions have to run all such scripts in all tabs and frames, and perform the check there, which means that every regular expression will be recompiled again and again without caching, thousands of times each day. The scripts may be very big, so megabytes of code will waste the CPU time and battery without being actually used.
> * MV3 userScripts API doesn't support smart blacklists offered by Violentmonkey, see the previous consideration minus the regular expression compilation.
> * MV3 userScripts API doesn't support setting GM_setValue separately, so we have to re-register the entire script and embed its storage data alongside the code, because GM_setValue/GM_getValue are synchronous and the data must be present at the moment the script runs. Re-registering is a complex operation inside the browser, so if we do it every time the script changes a value via these GM functions, the overhead will be staggering. We could throttle these updates, but only by a few milliseconds, otherwise when the script reloads the tab the old incorrect data will be loaded.
> * If we don't use the very limited MV3 userScripts API, we can't ensure the scripts will run at `document_start`:
>   
>   * "synchronous page mode" option won't be available as MV3 removes webRequestBlocking
>   * the background script will need 50ms - 500ms to start and load the scripts every time navigation occurs after a pause of 30 seconds, which may easily happen hundreds of times a day.

The developers have replied to the difficulties encountered with the new version, indicating that they have already tried upgrading to MV3 in private, they just don't want to incur the same extra CPU performance consumption as tampermonkey after the upgrade

**Comment by SekiBetu** [comment]
> > MV3 [...] it greatly improves security and efficiency.
> 
> It doesn't, those claims are greatly exaggerated to the point of being unsubstantiated propaganda.
> 
> * Nothing changes in regards to security for users of extensions that didn't do anything insecure.
> * Efficiency is lower for extensions like ours that need to process every navigation because MV3 disallows persistent background scripts, which means it will restart hundreds of times a day wasting resources much more than in MV2. No one in Chrome team has ever measured the impact, they just believe in an idealistic notion.
> * MV3 userScripts API doesn't support regular expressions, so extensions have to run all such scripts in all tabs and frames, and perform the check there, which means that every regular expression will be recompiled again and again without caching, thousands of times each day. The scripts may be very big, so megabytes of code will waste the CPU time and battery without being actually used.
> * MV3 userScripts API doesn't support smart blacklists offered by Violentmonkey, see the previous consideration minus the regular expression compilation.
> * MV3 userScripts API doesn't support setting GM_setValue separately, so we have to re-register the entire script and embed its storage data alongside the code, because GM_setValue/GM_getValue are synchronous and the data must be present at the moment the script runs. Re-registering is a complex operation inside the browser, so if we do it every time the script changes a value via these GM functions, the overhead will be staggering. We could throttle these updates, but only by a few milliseconds, otherwise when the script reloads the tab the old incorrect data will be loaded.
> * If we don't use the very limited MV3 userScripts API, we can't ensure the scripts will run at `document_start`:
>   
>   * "synchronous page mode" option won't be available as MV3 removes webRequestBlocking
>   * the background script will need 50ms - 500ms to start and load the scripts every time navigation occurs after a pause of 30 seconds, which may easily happen hundreds of times a day.

> I think we will make an MV3 version anyway because userScripts API offers some good features like world isolation and true document-start.

> Tampermonkey MV3 version has all the problems I've listed. But indeed, the power is practically the same, so @cyfung1031 is correct.

From what the above developers have revealed, I think the development team has already tried upgrading to MV3 privately, but just wasn't satisfied with the CPU performance consumption after the upgrade, so they never released a test version

**Comment by NeurekaSoftware** [comment]
Tampermonkey it is then...

**Comment by AziRizvi** [comment]
Implementing this has become the need of the hour because now chrome won't let you run this extension.

![Image](https://github.com/user-attachments/assets/82c5b979-8d52-464b-b515-236b40e54c57)

**Comment by AlexWayfer** [comment]
You can still press "Details" and then "Keep":

![Image](https://github.com/user-attachments/assets/c3a02658-aaa9-4e6f-b24f-37587172e42d)

**Comment by sitzmoebelchronograph** [comment]
Just a little note of things I explored today

1. in Google Chrome version 133 (stable version) I can enable it anyways
2. in Google Chrome version 134 (beta version) I can enable it anyways
3. in Google Chrome version 133 (unstable version) I can enable it anyways

@AlexWayfer mentioned a solid way, but there are also possibilities to install the Extention directly and not by Chrome Store.
The error itself like "manifest V2 is depricated" can be ignored as long as Google doesn't remove v2 (in June 2025 they wanna remove it)

Because of this is known for years I cant imagen the devs didn't solve it these days.....

**Comment by SekiBetu** [comment]
> Sy thank you to google for forcing ads all over the world......感谢谷歌在全世界投放广告......

The installed version of AdGuard still filters ads and web elements and supports some userscripts

**Comment by sitzmoebelchronograph** [comment]
> > Sy thank you to google for forcing ads all over the world......感谢谷歌在全世界投放广告......
> 
> The installed version of AdGuard still filters ads and web elements and supports some userscripts

until Juni this will work, but it's not filtering any kind of ads anymore :( but I guss this is a bit off topic here.

**Comment by SekiBetu** [comment]
> > > Sy thank you to google for forcing ads all over the world......感谢谷歌在全世界投放广告......
> > 
> > 
> > The installed version of AdGuard still filters ads and web elements and supports some userscripts
> 
> until Juni this will work, but it's not filtering any kind of ads anymore :( but I guss this is a bit off topic here.

I don't think so, the desktop version of adguard doesn't need to rely on a browser plugin to work properly with the MV3 standard

**Comment by sitzmoebelchronograph** [comment]
> > > > Sy thank you to google for forcing ads all over the world......感谢谷歌在全世界投放广告......
> > > 
> > > 
> > > The installed version of AdGuard still filters ads and web elements and supports some userscripts
> > 
> > 
> > until Juni this will work, but it's not filtering any kind of ads anymore :( but I guss this is a bit off topic here.
> 
> I don't think so, the desktop version of adguard doesn't need to rely on a browser plugin to work properly with the MV3 standard

yes, but should I install it with wine on my linux system? ADguard is now only helpful on macos and windows systems :/

**Comment by NeurekaSoftware** [comment]
> > > > > Sy thank you to google for forcing ads all over the world......感谢谷歌在全世界投放广告......
> > > > 
> > > > 
> > > > The installed version of AdGuard still filters ads and web elements and supports some userscripts
> > > 
> > > 
> > > until Juni this will work, but it's not filtering any kind of ads anymore :( but I guss this is a bit off topic here.
> > 
> > I don't think so, the desktop version of adguard doesn't need to rely on a browser plugin to work properly with the MV3 standard
> 
> yes, but should I install it with wine on my linux system? ADguard is now only helpful on macos and windows systems :/

AdGuard is now on Linux. See more: https://adguard.com/en/blog/adguard-for-linux-nightly.html

**Comment by knd775** [comment]
I think it would be extremely helpful for @tophf to let us know where they currently stand on this. I suspect we may just need to move on and let it die with MV2, but I really hope that's not the case.

**Comment by ELY3M** [comment]
Anyone ever forked this and updated it V3?

**Comment by NeurekaSoftware** [comment]
> I'd like to see this ported to MV3. Even as a savvy individual, the Tampermonkey interface is very overwhelming for me. I find this much easier to use.

I prefer the Violentmonkey UI too. That's why I am here and so passionate about the port.

The problem with somebody else doing the port is the project also needs to be maintained.

I hope it eventually gets ported but I am not holding my breath.

**Comment by Hecsall** [comment]
Joining the waitlist because apparently on the new Chrome 138.0.7204.101 you can't "force-enable" it anymore, its just permanently disabled (also tried with tricks in chrome://flags, nothing works)

<img width="719" height="217" alt="Image" src="https://github.com/user-attachments/assets/f3b23b7a-6a96-4974-8163-bae0ea7bb23d" />

**Comment by info1944** [comment]
i hope violentmonkey Dev team can figure this out, we love violentmonkey better than any other usercript extension

**Comment by Purfview** [comment]
Chrome is malware, a virus.

1) Backup Monkey's settings
2) Uninstall Chrome
3) Install Firefox
4) PROFIT

**Comment by selimhex** [comment]
hi @Purfview, i am sadly informing you of https://github.com/mozilla/bedrock/commit/d459addab846d8144b61939b7f4310eb80c5470e in case you haven't seen. Mozilla may not be that pure either.

anyways, i was just thinking of bumping this thread, because some discussions relating to this have been taking place in other issues, so thanks for doing that for me:)

what i see as positive news is that the maintainer of the project may have at least A plan on how to realise this:
@tophf https://github.com/violentmonkey/violentmonkey/issues/2341#issuecomment-3095426139
> Tentatively, my idea has been to use conditional compilation to produce both MV2 and MV3 versions out of the same source the way I did it in Stylus: see `MV3` [here](https://github.com/openstyles/stylus/blob/master/src/background/usercss-install-helper.js) and in [webpack.config.js](https://github.com/openstyles/stylus/blob/master/webpack.config.js#L490).

**Comment by Purfview** [comment]
> hi [@Purfview](https://github.com/Purfview), i am sadly informing you of [mozilla/bedrock@d459add](https://github.com/mozilla/bedrock/commit/d459addab846d8144b61939b7f4310eb80c5470e) in case you haven't seen. Mozilla may not be that pure either.

That's just ToS change, and look at a healthy reaction of the community. Anyway, Mozilla is not "pure" since it dropped Gecko.

Google actually sells your data and [allows trojans to be installed on your devices](https://github.com/Purfview/IMDb-Scout-Mod/issues/244#issuecomment-2513466194).
And what's reaction from the community? They want more trojans! 😆

**Comment by dnet890** [comment]
we have less 2 weeks left until Chromium v139 that does not support Manifest v2 reach stable and after that we have to say farewell to this extension. 

https://chromiumdash.appspot.com/schedule

**Comment by erosman** [comment]
If I may, there are a number of issues with the current UserScripts MV3 implementation on Chrome & Firefox:

- userScripts.execute() 	
  No support on Firefox at the moment
- Setting headers in GM xmlHttpRequest: Cookie, Host, Origin, Referer 	
  No support on Chrome at the moment
- Background GM xmlHttpRequest 	
  No support on Chrome at the moment
  >  In Manifest V3, XMLHttpRequest is not supported in background pages (provided by Service Workers). Please consider using its modern replacement, fetch().
  >  [Cross-origin XMLHttpRequest](https://developer.chrome.com/docs/extensions/mv3/xhr/)

- GM info.isIncognito 	
  No support at the moment
- GM registerMenuCommand, GM addValueChangeListener 	
  Require listener in the userScript context and no support at the moment

**Comment by divinity76** [comment]
Firefox has been corrupted. For a long time, they've been laying off developers and increasing C-suite (CEO, CTO, CFO) wages, and little else.

**Comment by selimhex** [comment]
uh, i just noticed this change on the violentmonkey docs:
https://github.com/violentmonkey/violentmonkey.github.io/commit/196814ab3cef929b14d3770e8a76d71dd385c4a1#diff-0305aa01e394ff1946b149dd8f794580cfaef6c4c08bc1532b497b3d94813b8cR31-R34

this doesn't sound good:/

**Comment by dnet890** [comment]
@selimhex Logically yes. Brave can do that. But, the script manager features are still barebones compared to violentmonkey. Besides, I heard some people want to fork this addon. But, idk if they can make it after chromium v139 stable that does not contained manifest v2 will arrive on August 5 .

**Comment by selimhex** [comment]
i don't have much confidence in that said forking tbh:/ i only saw that one issue where the person was talking about it, but they didn't seem to have done any work yet in this regard.

**Comment by Maartyl** [comment]
- I am using Brave when I need Chromium anyway.
- Any user able to check userscripts, probably does not want to use Chrome itself, unless they are forced to.
- It would only become a big issue if Chromium itself removed support for MV2 - for all browsers based on it. 

- In a sense, pushing people away from Chrome is a positive force. You might as well use another monkey if you are forced to use Chrome, which is not great, but I understand the complexity of MV3 rewrite; and it does not really further lower how much you can trust it..

- It's sad to hear about Mozilla, but let's hope it's not major, as I have seen mainly good stuff.

@dnet890  Brave can run Violentmonkey itself. - If that stops working, that would be sad, though.

**Comment by Yay295** [comment]
imo a good thing to do would be to update the Chrome version to MV3 but just as a read-only version, so that people can at least access the scripts they have installed.

**Comment by dnet890** [comment]
@Maartyl it is not separate ViolentMonkey addon. Brave have their own build in script manager (https://brave.com/privacy-updates/32-custom-scriptlets/). But, this feature is just barebones compared to ViolentMonkey.
@selimhex in this digital era especially for security reason. it is crazy to stop update your own browser just for Manifest v2.

**Comment by divinity76** [comment]
How about we stop this unproductive quarrelling and actually start coding?

**Comment by divinity76** [comment]
Do we actually have a testsuite for violentmonkey? I see there's a CI, but it seems to just test that the extension builds correctly

**Comment by andrewdbond** [comment]
> Brave

Tampermonkey ( https://www.tampermonkey.net ) seems like the more obvious replacement as it works across Chrome (v3), Firefox, Safari, Brave/Edge/etc. and has _about_ as many features as Violentmonkey including cross-browser sync, and nice items like the ability to diff userscript updates -- and I say that as a long-time Chrome Violentmonkey user. (Sadly, Tampermonkey is not open source.)

**Comment by cyfung1031** [comment]
> > Brave
> 
> Tampermonkey ( https://www.tampermonkey.net ) seems like the more obvious replacement as it works across Chrome (v3), Firefox, Safari, Brave/Edge/etc. and has about as many features as Violentmonkey including cross-browser sync, and nice items like the ability to diff userscript updates -- and I say that as a long-time Chrome Violentmonkey user. (Sadly, Tampermonkey is not open source.)

Already mentioned [in another issue](https://github.com/violentmonkey/violentmonkey/issues/2284#issuecomment-3010784673) that there is alternative to Violentmonkey for MV3 which is open source.

**Comment by MikeSoftZP** [comment]
temporary solution to run Violentmonkey on Chrome 140+ (tested on 140.0.7339.81 [macos]):
run Chrome with parameters: `--disable-features=ExtensionManifestV2Unsupported,ExtensionManifestV2Disabled`

`/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --disable-features=ExtensionManifestV2Unsupported,ExtensionManifestV2Disabled`

also, you could create new shortcut via Automator.app:
New -> Application -> Run shell script -> `open -a "Google Chrome.app" --args --disable-features=ExtensionManifestV2Unsupported,ExtensionManifestV2Disabled` -> Save

**Comment by karlhorky** [comment]
> @tophf in [comment 3095426139](https://github.com/violentmonkey/violentmonkey/issues/2341#issuecomment-3095426139) (21 Jul 2025): Tentatively, my idea has been to use conditional compilation to produce both MV2 and MV3 versions out of the same source the way I did it in Stylus: see `MV3` [here](https://github.com/openstyles/stylus/blob/master/src/background/usercss-install-helper.js?rgh-link-date=2025-07-21T06%3A36%3A00.000Z) and in [webpack.config.js](https://github.com/openstyles/stylus/blob/master/webpack.config.js?rgh-link-date=2025-07-21T06%3A36%3A00.000Z#L490).

@tophf should we start an actionable TODO checklist in the issue description at the top of the page? Something where contributors can also try to help out here?

Eg. maybe something like the following?

(I'm not sure I fully understood your plan and I'm not super familiar with Violentmonkey internals, so it's possible this is partly inaccurate, so feel free to change or discard)

---

Build system: flags and conditionals
- [ ] Add build flags in `package.json` (eg. `build:mv2`, `build:mv3` with `V3=1`)
- [ ] Define `const MV3 = !!process.env.V3` in `webpack.config.js`
- [ ] Inject a global `__MV3__` via `DefinePlugin` for conditional code

Manifest changes (MV3 specifics)
- [ ] Set `"manifest_version": 3`
- [ ] Use `"background": { "service_worker": "sw.js", "type": "module" }`
- [ ] Replace `browser_action` / `page_action` with `action`
- [ ] Move host patterns from `permissions` to `host_permissions`
- [ ] Update `web_accessible_resources` to array-of-objects (`resources` + `matches`)
- [ ] Convert `content_security_policy` to object form under `extension_pages`

Generate manifest at build time
- [ ] Add base `manifest.json` plus `manifest.v3.json` overrides
- [ ] Use `copy-webpack-plugin` with `transform` to merge overrides when `MV3`

Background script → Service Worker (MV3)
- [ ] Identify current background entry point
- [ ] Switch MV3 build to `sw.js` (service worker) entry
- [ ] Remove any `window`/DOM usage from background
- [ ] Replace `XMLHttpRequest` with `fetch`
- [ ] Create `offscreen.html` for tasks needing DOM/audio/crypto parsing
- [ ] Add `offscreen.js` as a separate Webpack entry
- [ ] Implement SW logic to create the offscreen document and message with it

In-memory globals and timers → persisted storage and alarms
- [ ] Persist any global state (`chrome.storage.local` or `chrome.storage.session`; consider IndexedDB for large data)
- [ ] Refactor storage access to async/await (no sync reads)
- [ ] Rehydrate required state on SW start (initialize from storage when needed)
- [ ] Replace `setTimeout` / `setInterval` with `chrome.alarms` for delayed/periodic work
- [ ] Ensure tasks are event-driven so SW can suspend safely (no keep-alive loops)

Network interception and script execution updates
- [ ] Audit all `chrome.webRequest` listeners
- [ ] Migrate request modification to `chrome.declarativeNetRequest` (static/dynamic rulesets)
- [ ] Replace `chrome.tabs.executeScript` with `chrome.scripting.executeScript` (add `"scripting"` permission)
- [ ] Replace `chrome.browserAction` / `chrome.pageAction` calls with `chrome.action`
- [ ] Review code-evaluation patterns for MV3 compliance (no remote `eval`; fetch and inject via `scripting` or content scripts)

**Comment by MikeSoftZP** [comment]
> Sir pls help, how this make on windows PC

create shortcut for chrome.exe, open its properties.
add arguments in 'target' field, right after full path (after quotes).
so it will look like this: 
`"C:\Program Files\Google\Chrome\Application\chrome.exe" --disable-features=ExtensionManifestV2Unsupported,ExtensionManifestV2Disabled`

**Comment by online** [comment]
@MikeSoftZP, Please tell me how to solve this problem for macOS?

**Comment by injust** [comment]
Could we please not turn this thread into a Chrome helpdesk?

**Comment by krystian3w** [comment]
They probably knew that.

**Comment by divinity76** [comment]
Now that TamperMonkey has gone open-source, it's an excellent alternative to ViolentMonkey on MV3 browsers.

**Comment by divinity76** [comment]
> You made me lose 15 minutes checking this incorrect info. Next time, add a wink smiley.

They really have: https://github.com/Tampermonkey/tampermonkey

But it must have been very recent: There is no link to their github repo from their official website, nor in the TamperMonkey extension itself, as far as I can see. The "Contribute" section of their website is just a monetary donation page, and the GitHub link in the extension itself links to the main developer's personal github page, not the tampermonkey github page.

But the repo itself is legit, as the main developer is active on that github project's issue tracker.

**Comment by injust** [comment]
> They really have: https://github.com/Tampermonkey/tampermonkey

Have you considered reading the first line of the README?

**Comment by MetapherLab** [comment]
BTW Violentmonkey is still running on the lastest Chrome (144.0.7559.97), and will be continue working at least half I year AFAIK.

This is what I did (AI-supported):

> To keep Manifest V2 extensions working in Chrome 144, you need two things.
> 
> First, launch Chrome with the following flags:
> --disable-features=ExtensionManifestV2Unsupported,ExtensionManifestV2Disabled
> 
> Second, set the required Chrome policy in the Windows Registry:
> HKEY_LOCAL_MACHINE\SOFTWARE\Policies\Google\Chrome
> ExtensionManifestV2Availability (DWORD) = 2
> This forces Chrome to fully allow Manifest V2 extensions.
> 
> Optional steps if Chrome still blocks extensions:
> 
> Add an allowlist via ExtensionInstallAllowlist (set “*” or specific extension IDs).
> Make sure ExtensionInstallBlocklist is empty.
> Check chrome://policy to verify that Chrome is actually loading your policies and that no conflicting policies are active.
> 
> With these settings, Manifest V2 extensions continue to work even in newer Chrome builds like 144.

**Comment by Prigoryan** [comment]
@MetapherLab another option is to use https://github.com/scriptscat/scriptcat, which is basically the same as violentmonkey and works with MV3

**Comment by gunir** [comment]
> [@MetapherLab](https://github.com/MetapherLab) another option is to use https://github.com/scriptscat/scriptcat, which is basically the same as violentmonkey and works with MV3

Saying same is a bit over statement, even Tampermonkey MV3 is breaking a lot of scripts, and then haivng to enable Developer Mode is something like 1% userbase can, also overstatement because realistically less than 1%

**Comment by PPPDUD** [comment]
@tophf Are you going to begin work on this? `chrome.UserScripts` is mature now.

**Comment by marbens-arch** [comment]
@tophf Hasn't this issue been resolved too?

**Comment by marbens-arch** [comment]
@electrotype The context is that Violentmonkey supports MV3 since v2.43.0. #2284 was closed, but this issue wasn't. Looks like an oversight to me.

**Comment by karlhorky** [comment]
@tophf thanks for implementing MV3! 🎉

Appreciate you putting in the time and effort for this ❤️ 🐒 Violentmonkey is great.

Even though I and others personally think MV3 was a net negative, user-hostile new addition to Chrome, it's part of the requirements now, so it's important to implement.

**Comment by KaKi87** [comment]
> > MV3 [...] it greatly improves security and efficiency.
> 
> It doesn't, those claims are greatly exaggerated to the point of being unsubstantiated propaganda.
> 
>     * Nothing changes in regards to security for users of extensions that didn't do anything insecure.
> 
>     * Efficiency is lower for extensions like ours that need to process every navigation because MV3 disallows persistent background scripts, which means it will restart hundreds of times a day wasting resources much more than in MV2. No one in Chrome team has ever measured the impact, they just believe in an idealistic notion.
> 
>     * MV3 userScripts API doesn't support regular expressions, so extensions have to run all such scripts in all tabs and frames, and perform the check there, which means that every regular expression will be recompiled again and again without caching, thousands of times each day. The scripts may be very big, so megabytes of code will waste the CPU time and battery without being actually used.
> 
>     * MV3 userScripts API doesn't support smart blacklists offered by Violentmonkey, see the previous consideration minus the regular expression compilation.
> 
>     * MV3 userScripts API doesn't support setting GM_setValue separately, so we have to re-register the entire script and embed its storage data alongside the code, because GM_setValue/GM_getValue are synchronous and the data must be present at the moment the script runs. Re-registering is a complex operation inside the browser, so if we do it every time the script changes a value via these GM functions, the overhead will be staggering. We could throttle these updates, but only by a few milliseconds, otherwise when the script reloads the tab the old incorrect data will be loaded.
> 
>     * If we don't use the very limited MV3 userScripts API, we can't ensure the scripts will run at `document_start`:
>       
>       * "synchronous page mode" option won't be available as MV3 removes webRequestBlocking
>       * the background script will need 50ms - 500ms to start and load the scripts every time navigation occurs after a pause of 30 seconds, which may easily happen hundreds of times a day.

What about all this ?

**Comment by tophf** [comment]
Chromium didn't fix any of those, I guess their main developer who decides what to do lives in their own cognitive bubble. Violentmonkey has to use this new userScripts API in an ineffective mode practically the same way it works in MV2, i.e. instead of registering the scripts and letting the browser match them it matches them explicitly and injects into the page. On the bright side I've managed to utilize the API's registration machinery for this pre-matched array of scripts so there's an unexpected improvement `document-start` in MV3 is now real for most webpages even without synchronous/instant inject option.

**Comment by jesus2099** [comment]
> https://github.com/violentmonkey/violentmonkey/issues/1934#issuecomment-2154174223
> MV3 userScripts API doesn't support regular expressions, so extensions have to run all such scripts in all tabs and frames, and perform the check there

@tophf, is this still true?

Are we talking about the `//@include /^https?://((beta|test)\.)?musicbrainz\.(org|eu)/this that/` kind of metadata block entries?

How to improve behaviour with our userscripts? 
Replace regex includes by many basic includes or Chrome matches?

**Comment by cyfung1031** [comment]
> > [#1934 (comment)](https://github.com/violentmonkey/violentmonkey/issues/1934#issuecomment-2154174223)
> > MV3 userScripts API doesn't support regular expressions, so extensions have to run all such scripts in all tabs and frames, and perform the check there
> 
> [@tophf](https://github.com/tophf), is this still true?
> 
> Are we talking about the `//@include /^https?://((beta|test)\.)?musicbrainz\.(org|eu)/this that/` kind of metadata block entries?
> 
> How to improve behaviour with or userscripts? Replace regex includes by many basic includes or Chrome matches?

I believe @tophf was taking about the internal mechanism.
The API itself does not support regular expressions, so VM have to run all such scripts then check the regular expressions using JavaScript coding rather than API filtering.

**Comment by KaKi87** [comment]
> The API itself does not support regular expressions, so VM have to run all such scripts then check the regular expressions using JavaScript coding rather than API filtering.

But the API requires a finite list of qualified domain names, right ? That cannot be reliably geneated from a RegExp, which means even if `@include` would remain and be computed by ViolentMonkey, a `@match` would still be required.

**Comment by tophf** [comment]
To clarify, Violentmonkey MV3 doesn't register individual scripts. It registers an array of scripts that match a navigated URL using its own matching mechanism that supports regexps and AFAICT this registration is fast enough to occur at a real document-start meaning it's used right away without waiting for asynchronous communication which is necessary when synchronous/instant inject option is not enabled. Currently there's no need to change anything in your scripts although simplifying regexps is always a good thing if you can realistically do it. In the future I'll probably add individual registration of scripts although I abhor the idea of people needlessly having tons of non-cacheable code from scripts that use regexps run in each webpage and frame thousands of times a day so maybe I'll wait for chromium devs to implement at least the simplified RE2 syntax for regexps first.

**Comment by KaKi87** [comment]
So I immediately noticed it's much slower : my [Dark Reader dynamic blacklist](https://git.kaki87.net/KaKi87/userscripts/src/branch/master/darkReaderDynamicBlacklist) (which uses `@run-at document-start`) now flickers instead of running smoothly.

> their main developer who decides what to do lives in their own cognitive bubble. Violentmonkey has to use this new userScripts API

Regarding this, to be honest I didn't think you'd cave, rather keep supporting browsers that keep supporting MV2 and drop the rest.

I'm gonna downgrade.

**Comment by tophf** [comment]
In all my tests `document-start` became much faster so I guess the problem you've encountered is caused by the `document.head` check in your script: previously `document-start` was slower and ran the scripts when `document.head` already existed so your script immediately started its work, but now there's no `document.head` yet so your script waits for DOMContentLoaded. Usually the proper solution is to use `document.documentElement` as it can host any DOM nodes directly just fine if you use appendChild or insertBefore and not the new method like append(). Or you can use MutationObserver to wait for document.head which usually will occur in a couple of milliseconds effectively replicating the previous slower timing of document-start.

**Comment by KaKi87** [comment]
Well, you were right, there was no `document.head` yet, so I'm now waiting for it, and the flickering is gone. I'm not using `document.documentElement` though because that's not where Dark Reader expects the `darkreader-lock` element to be added.

However, I now see that installing userscripts makes the browser's file download prompt appear at the same time as Violentmonkey's install page, so I have to dismiss the former before being able to interact the former, all the time, and that's annoying, an MV3 consequence I assume.

**Comment by tophf** [comment]
The prompt is removed in Chrome 152. For older browsers I've fixed the suppression rule which missed some cases.

**Comment by KaKi87** [comment]
I'm on Vivaldi 8.0, which is the last version to support MV2, so I'll be stuck there.

**Comment by yeezylife** [comment]
> > MV3 [...] it greatly improves security and efficiency.
> 
> It doesn't, those claims are greatly exaggerated to the point of being unsubstantiated propaganda.
> 
> * Nothing changes in regards to security for users of extensions that didn't do anything insecure.
> * Efficiency is lower for extensions like ours that need to process every navigation because MV3 disallows persistent background scripts, which means it will restart hundreds of times a day wasting resources much more than in MV2. No one in Chrome team has ever measured the impact, they just believe in an idealistic notion.
> * MV3 userScripts API doesn't support regular expressions, so extensions have to run all such scripts in all tabs and frames, and perform the check there, which means that every regular expression will be recompiled again and again without caching, thousands of times each day. The scripts may be very big, so megabytes of code will waste the CPU time and battery without being actually used.
> * MV3 userScripts API doesn't support smart blacklists offered by Violentmonkey, see the previous consideration minus the regular expression compilation.
> * MV3 userScripts API doesn't support setting GM_setValue separately, so we have to re-register the entire script and embed its storage data alongside the code, because GM_setValue/GM_getValue are synchronous and the data must be present at the moment the script runs. Re-registering is a complex operation inside the browser, so if we do it every time the script changes a value via these GM functions, the overhead will be staggering. We could throttle these updates, but only by a few milliseconds, otherwise when the script reloads the tab the old incorrect data will be loaded.
> * If we don't use the very limited MV3 userScripts API, we can't ensure the scripts will run at :
>   `document_start`
>   * "synchronous page mode" option won't be available as MV3 removes webRequestBlocking
>   * the background script will need 50ms - 500ms to start and load the scripts every time navigation occurs after a pause of 30 seconds, which may easily happen hundreds of times a day.

Does these problems still exist in violentmonkey beta mv3? should I keep using mv2 version for better performance? or mv3 version is good enough today performance wise

**Comment by tophf** [comment]
@yeezylife, the API itself is still as bad as it was, but I've tried to mitigate the problems, so for most users MV3 version should be comparable in performance. It also provides a faster `document-start` injection mode `[x] Alternative page mode` in the options but it should be used only if your scripts and their data are small e.g. mine are less than 200kB per webpage.
